### PR TITLE
GRAN DAG

### LINKED ISSUE #3405: [ENH] GRAN-DAG Algorithm
**Is your feature request related to a problem? Please describe.**
 GraN-DAG (Gradient-based Neural DAG Learning) parameterizes each node’s conditional distribution using a neural network that takes all other variables as input. A continuous differentiable acyclicity constraint (adapted from NOTEARS) is imposed so that the entire structure learning problem can be solved end-to-end with gradient descent. A Lagrangian augmentation scheme is used to enforce the acyclicity constraint as a hard constraint at convergence. Unlike linear NOTEARS, GraN-DAG captures non-linear relationships without requiring explicit functional form assumptions.

**Additional context**
Paper: [Gradient-Based Neural DAG Learning](https://arxiv.org/abs/1906.02226)
Reference codebase: [kurowasan/GraN-DAG](https://github.com/kurowasan/GraN-DAG)
Related : #2931, pgmpy/enhancement_proposals#4

Note that this is part of an ongoing mentorship project

### PR DESCRIPTION
**Pull request by Manas-7854**
**The following checklist is mandatory**

Your PR will be closed if you remove the checklist. Use of LLMs is **strictly forbidden** for any part of this checklist (including for improving language), and will result in a **ban** if we find any use of LLMs.

### Your checklist for this pull request

- [.] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/CONTRIBUTING.md)?
- [] Does the PR fully address the linked issue and is within its defined scope? If you are still working on the PR, mark it as draft. - No this is a draft PR
- [ ] Are all the GitHub Actions checks passing? If not, mark your PR as draft while you fix it.- marked as draft

If you have used AI/LLMs for any assistance, please answer the following questions. Please refer [#2622](https://github.com/pgmpy/pgmpy/pull/2622) for an example of the level of detail we expect:

- [.] Have you reviewed our [AI usage policy](https://github.com/pgmpy/pgmpy/blob/dev/CONTRIBUTING.md#ai-usage-policy)?
- [.] Are you able to fully explain your changes? We expect you to fully understand the algorithm and take full responsibility for any changes in this PR.

- Please list the AI tool(s) used, along with the model and its version used.
- Claude Code

- Please describe in **detail** how and what you used AI assistance for? Please outline your whole workflow with the AI tool.

- Claude Code - given the API used it to write function signature and `__init__`

- What steps have you taken to verify that the changes correctly address the issue? What edge cases have you considered? Other than running tests, what else have you verified?

TBD

- Have you used AI for generating tests? Can you compress them into a smaller number of tests without losing coverage?

TBD

### Issue number(s) that this pull request fixes
- Fixes #3405, #3408 

### List of changes to the codebase in this pull request
- Implements GRAN DAG algorithm inside `pgmpy/causal_discovery/grandag.py`
- Implements tests for GRAN DAG algorithm inside `pgmpy/tests/test_causal_discovery/test_grandag.py`

### CODE REVIEW

File: pgmpy/causal_discovery/gran_dag.py [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+    W.fill_diagonal_(0)
+    W[W < edge_threshold] = 0
+
+    A = W > 0
+    if _is_acyclic(A):
+        return A.to(J.dtype)
+
+    ts = torch.unique(W[W > 0])  # ascending
+    lo, hi = 0, len(ts) - 1  # invariant: A(ts[hi]) is acyclic
+    EPS = 1e-8
```
Comment (ankurankan) [inline pgmpy/causal_discovery/gran_dag.py]: Should be an argument.
